test/jdk/java/lang/Thread/virtual/MonitorPinnedEvents.java

```diff
@@ -0,0 +1,478 @@
+     * @param timed true for a timed-wait, false for an untimed-wait
+     * @param interrupt true to interrupt the thread, false to notify
+     */
+    private void testObjectWait2(boolean timed, boolean interrupt) throws Exception {
+        try (Recording recording = new Recording()) {
+            recording.enable("jdk.VirtualThreadPinned");
+            recording.start();
+
+            // the thread that acquires the monitor after the main thread releases it
+            var nextOwner = new AtomicReference<Thread>();
+
+            Thread vthread;
+            try {
+                Object lock = new Object();
+
+                // start virtual that waits in Object.wait while pinned
+                var ready1 = new AtomicBoolean();
+                vthread = Thread.ofVirtual().start(() -> {
+                    VThreadPinner.runPinned(() -> {
+                        boolean notify = !interrupt;
+                        synchronized (lock) {
+                            ready1.set(true);
+                            try {
+                                // wait while pinned
+                                if (timed) {
+                                    lock.wait(Long.MAX_VALUE);
+                                } else {
+                                    lock.wait();
+                                }
+                                if (notify) {
+                                    nextOwner.compareAndSet(null, Thread.currentThread());
+                                }
+                            } catch (InterruptedException e) {
+                                if (interrupt) {
+                                    nextOwner.compareAndSet(null, Thread.currentThread());
+                                }
+                            }
+                        }
+                    });
+                });
+                await(ready1, vthread, timed ? Thread.State.TIMED_WAITING : Thread.State.WAITING);
+
+                // platform thread that blocks on monitor enter
+                Thread thread2;
+                synchronized (lock) {
+                    thread2 = Thread.ofPlatform().start(() -> {
+                        synchronized (lock) {   // <--- blocks here
+                            if (nextOwner.compareAndSet(null, Thread.currentThread())) {
+                                // delayed release of monitor if acquired before virtual thread
+                                try {
+                                    Thread.sleep(DELAY);
+                                } catch (InterruptedException e) {
+                                }
+                            }
+                        }
+                    });
+
+                    // interrupt/notify and release monitor to allow one of the threads to acquire
+                    if (interrupt) {
+                        vthread.interrupt();
+                    } else {
+                        synchronized (lock) {
```

Comment:
Not sure if this was intentional but we already own the monitor.

Reply:
Well spotted, it was copied down from the other test other test that can also interrupt or notify, this main thread in this case already owns the monitor so no need to enter again. Thanks!